### ISSUE TITLE
docker daemon panic with swarmkit and lots of services(300+)

### ISSUE THREAD

**Issue by wrfly** [docker/docker#28465]
10+ worker nodes and 3 manager nodes

### docker info:
```
Containers: 209
 Running: 7
 Paused: 0
 Stopped: 202
Images: 8
Server Version: 1.12.3
Storage Driver: devicemapper
 Pool Name: docker-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 7.44 GB
 Data Space Total: 42.9 GB
 Data Space Available: 35.46 GB
 Metadata Space Used: 9.806 MB
 Metadata Space Total: 109.1 MB
 Metadata Space Available: 99.25 MB
 Thin Pool Minimum Free Space: 4.29 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local rexray
 Network: host bridge null overlay
Swarm: active
 NodeID: 22vdutsft30ybydl1sb21237r
 Is Manager: false
 Node Address: 10.10.29.24
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.64 GiB
Name: 10-10-29-24
ID: DN2L:4CMR:U5S5:XAOT:7BSZ:SZXL:FNVI:GQSR:3Q4K:KJSS:PB7D:5PGQ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

### logs:
```
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: time="2016-11-16T11:18:12+08:00" level=error msg="setting up rule failed,
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: panic: runtime error: invalid memory address or nil pointer dereference
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: [signal 0xb code=0x1 addr=0x0 pc=0xde171e]
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: goroutine 9610 [running]:
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: panic(0x1a90b60, 0xc82000a080)
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: /usr/local/go/src/runtime/panic.go:481 +0x3e6
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: github.com/miekg/dns.getUDPSocketName(0x0, 0x0, 0x0, 0x0, 0x0)
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/miekg/dns/udp_li
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: github.com/miekg/dns.setUDPSocketOptions(0x0, 0x0, 0x0)
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/miekg/dns/udp.go
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: github.com/miekg/dns.(*Server).ActivateAndServe(0xc82087e1a0, 0x0, 0x0)
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/miekg/dns/server
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: github.com/docker/libnetwork.(*resolver).Start.func1(0xc82087e1a0)
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/docker/libnetwor
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: created by github.com/docker/libnetwork.(*resolver).Start
Nov 16 11:18:12 10-10-135-223 dockerd[20196]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/docker/libnetwor
Nov 16 11:18:15 10-10-135-223 dockerd[20196]: time="2016-11-16T11:18:14+08:00" level=error msg="setting up rule failed,
Nov 16 11:18:14 10-10-135-223 systemd[1]: docker.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Nov 16 11:18:15 10-10-135-223 systemd[1]: Unit docker.service entered failed state.
Nov 16 11:18:15 10-10-135-223 systemd[1]: docker.service failed.
```

**Comment by mavenugo** [comment]
@sanimej can you PTAL at this ?

**Comment by wrfly** [comment]
# Panic Again on several nodes... sad...

![1](https://cloud.githubusercontent.com/assets/7827389/20381269/b3e24a62-ace1-11e6-9f13-a459c2ab9d28.png)

![2](https://cloud.githubusercontent.com/assets/7827389/20381274/b7e12476-ace1-11e6-9592-c28421d0b5ca.png)

![3](https://cloud.githubusercontent.com/assets/7827389/20381279/bb5d54ee-ace1-11e6-86d3-b36f79b7c6f4.png)

**Comment by sanimej** [comment]
This panic was reported in #28112 as well.

@wrfly From the decode the only possibility I can think of is: a container comes up and while the embedded DNS server is being setup the container goes down and the embedded DNS sever gets shutdown. In your setup do you think its possible for some containers to come up and go down very rapidly ? Have you noticed any specific trigger for this panic ?

**Comment by wrfly** [comment]
Here are more logs:

```
Nov 18 18:37:24 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:24+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:28 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:28+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:32 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:32+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:36 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:36+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:40 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:40+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:44 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:44+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:48 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:48+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:52 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:52+08:00" level=info msg="Firewalld running: false"
Nov 18 18:37:56 10-10-237-17 dockerd[855]: time="2016-11-18T18:37:56+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:00 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:00+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:04 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:04.425310056+08:00" level=warning msg="failed to create proxy for port 60370: listen tcp :60370: bind: a
Nov 18 18:38:04 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:04+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:07 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:07+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:11 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:11.566660548+08:00" level=warning msg="failed to create proxy for port 60169: listen tcp :60169: bind: a
Nov 18 18:38:11 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:11+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:11 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:11.756244246+08:00" level=error msg="containerd: start container" error="containerd: container did not s
Nov 18 18:38:11 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:11.763978675+08:00" level=error msg="Create container failed with error: containerd: container did not s
Nov 18 18:38:12 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:12.102412728+08:00" level=error msg="fatal task error" error= module=taskmanager task.id=9zmt8qvh9hlzu6l
Nov 18 18:38:12 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:12.289088548+08:00" level=error msg="container status unavailable" error="context canceled" module=taskm
Nov 18 18:38:15 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:15+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:18 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:18+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:22 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:22.832268931+08:00" level=warning msg="failed to create proxy for port 60330: listen tcp :60330: bind: a
Nov 18 18:38:22 10-10-237-17 dockerd[855]: time="2016-11-18T18:38:22+08:00" level=info msg="Firewalld running: false"
Nov 18 18:38:22 10-10-237-17 dockerd[855]: panic: runtime error: invalid memory address or nil pointer dereference
Nov 18 18:38:22 10-10-237-17 dockerd[855]: [signal 0xb code=0x1 addr=0x0 pc=0xde171e]
Nov 18 18:38:22 10-10-237-17 dockerd[855]: goroutine 5517 [running]:
Nov 18 18:38:22 10-10-237-17 dockerd[855]: panic(0x1a90b60, 0xc82000a080)
Nov 18 18:38:22 10-10-237-17 dockerd[855]: /usr/local/go/src/runtime/panic.go:481 +0x3e6
Nov 18 18:38:22 10-10-237-17 dockerd[855]: github.com/miekg/dns.getUDPSocketName(0x0, 0x0, 0x0, 0x0, 0x0)
Nov 18 18:38:22 10-10-237-17 dockerd[855]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/miekg/dns/udp_linux.go:68 +0xce
Nov 18 18:38:22 10-10-237-17 dockerd[855]: github.com/miekg/dns.setUDPSocketOptions(0x0, 0x0, 0x0)
Nov 18 18:38:22 10-10-237-17 dockerd[855]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/miekg/dns/udp.go:23 +0x37
Nov 18 18:38:22 10-10-237-17 dockerd[855]: github.com/miekg/dns.(*Server).ActivateAndServe(0xc8213f7e10, 0x0, 0x0)
Nov 18 18:38:22 10-10-237-17 dockerd[855]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/miekg/dns/server.go:349 +0x1b7
Nov 18 18:38:22 10-10-237-17 dockerd[855]: github.com/docker/libnetwork.(*resolver).Start.func1(0xc8213f7e10)
Nov 18 18:38:22 10-10-237-17 dockerd[855]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/docker/libnetwork/resolver.go:122 +0x21
Nov 18 18:38:22 10-10-237-17 dockerd[855]: created by github.com/docker/libnetwork.(*resolver).Start
Nov 18 18:38:22 10-10-237-17 dockerd[855]: /root/rpmbuild/BUILD/docker-engine/vendor/src/github.com/docker/libnetwork/resolver.go:123 +0x2b1
Nov 18 18:38:22 10-10-237-17 systemd[1]: docker.service: main process exited, code=exited, status=2/INVALIDARGUMENT
Nov 18 18:38:22 10-10-237-17 systemd[1]: Unit docker.service entered failed state.
Nov 18 18:38:22 10-10-237-17 systemd[1]: docker.service failed.
```

It happends again, and I have no idea what triggered this.

**Comment by sanimej** [comment]
@wrfly Identified a possible sequence in which this panic can happen. Working on a fix..

**Comment by wrfly** [comment]
@sanimej  Does anything related to iptables? I got an error when a service down:
![screenshot from 2016-11-19 19-55-44](https://cloud.githubusercontent.com/assets/7827389/20455314/082c1766-ae94-11e6-9481-11e520ca9d98.png)

**Comment by wrfly** [comment]
@sanimej Not very rapidly but in some cases, the container can be shutdown in seconds. When I create a service I also get an error that message is "Address is already used", it happens just few times.

**Comment by wrfly** [comment]
@aboch Would you mind tell me why this happened? Because I stuck in this for days and i'm really confused. Thank you.

**Comment by aboch** [comment]
@wrfly The code which starts/stops the embedded DNS server was not handling concurrent start/stop calls well. @sanimej identified and fixed the issue, he can add more.

**Comment by sanimej** [comment]
@wrfly From analyzing the stack trace of the panic, the reason seems to be because of a container coming up and going down very quickly. This lead to a nil pointer being accessed in the code path that sets up the embedded DNS. [libnetwork PR](https://github.com/docker/libnetwork/pull/1561) has some more details if you are interested. This fix will be available in 1.13 RC2.
